src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,397 @@
+        # This implies that it's possible for values to be
+        # quite stale in bizarre failure cases (value not read
+        # for a long time, then updated in cluster by some other
+        # instance, then cluster dies, then this instance reads,
+        # gets the old value)
+        if self._ioctx:
+            try:
+                with rados.ReadOpCtx() as op:
+                    omap_iter, ret = self._ioctx.get_omap_vals_by_keys(
+                        op, (key,)
+                    )
+                    assert ret == 0  # ???
+                    # TODO: does this need to be async?
+                    self._ioctx.operate_read_op(op, "kvstore")
+                    kv = dict(omap_iter)
+                    if key in kv:
+                        self._db[key] = kv[key]
+                    else:
+                        # key not present in cluster kvstore, make sure
+                        # it's also not present in db (prevent stale cache)
+                        if key in self._db:
+                            del self._db[key]
+
+            except Exception as e:
+                # TODO: deal with this (log it? ignore it?)
+                logger.error(str(e))
+
+        value = self._db.get(key)
         if not value:
             return None
         return value.decode("utf-8")
 
-    async def get_prefix(self, key: str) -> List[str]:
+    async def get(self, key: str) -> Optional[str]:
+        """Get value for provided key"""
+        # Wraps _get() so we can use _get() outside asyncio in _config_notify()
+        return self._get(key)
+
+    async def get_prefix(self, key_prefix: str) -> List[str]:
         """Get a range of keys with a prefix"""
-        assert self._client
+        # This is ugly, because pulling it out of omap is nice (there's a prefix
+        # arg), but getting it from dbm means iterating through everything.
+        # On the plus side, we're probably not talking about a lot of data...
+        # But there's a perf optimization argument here for taking the values from
+        # omap, setting them in the dbm, and returning the values from *omap* if
+        # present, then only falling back to iterating dbm if we don't have the
+        # values, rather than using the simpler logic in get().  But that might be
+        # premature optimization, so possibly more straightforward to do the
+        # simple implementation, then comment this for later use.
         values = []
-        async for value, _ in self._client.get_prefix(key):  # type: ignore[attr-defined]
-            values.append(value.decode("utf-8"))
+
+        if self._ioctx:
+            try:
+                with rados.ReadOpCtx() as op:
+                    omap_iter, ret = self._ioctx.get_omap_vals(
+                        op,
+                        start_after="",
+                        filter_prefix=key_prefix,
+                        max_return=1000,
+                    )
```

Comment:
Have you run this through black? I have a feeling this needs to be reformatted.

Reply:
Black is passing according to github checks.

Reply:
Black actually told me to do that :-)